### PR TITLE
Only send headers if sessions changed

### LINKED ISSUE #36: Unchanged sessions should not be persisted on every requests
- [x] I was not able to find an [open](https://github.com/zendframework/zend-expressive-session/issues?q=is%3Aopen) or [closed](https://github.com/zendframework/zend-expressive-session/issues?q=is%3Aclosed) issue matching what I'm seeing.
 - [x] This is not a question. (Questions should be asked on [slack](https://zendframework.slack.com/) ([Signup for Slack here](https://zendframework-slack.herokuapp.com/)) or our [forums](https://discourse.zendframework.com/).)

Provide a narrative description of what you are trying to accomplish.

### Code to reproduce the issue

Follow the documentation to pipe the middleware early for the whole application in `config/pipeline.php` like so:

```php
$app->pipe(\Zend\Expressive\Session\SessionMiddleware::class);
$app->pipeRoutingMiddleware();
```

And configure the session persistence implementation with `zendframework/zend-expressive-session-ext` (the only one published by Zend AFAIK) in `config/config.php`:

```php
 $aggregator = new ConfigAggregator([
   \Zend\Expressive\Session\Ext\ConfigProvider::class,
    // [...]
], $cacheConfig['config_cache_path']);

return $aggregator->getMergedConfig();
```

### Expected results

I expect to see a `Set-Cookie` header only when I write to the session (typically when a user logs in). And I expect to never see a `Set-Cookie` header if the session is not written to (typically most other requests after a login happened).

### Actual results

Every single request has the `Set-Cookie` header. This has several downsides.

First, it's of course a (very) slight performance hit (on server CPU, client CPU and network transfer) for no gain at all.

Second, the default behavior of `\Zend\Expressive\Session\Ext\PhpSessionPersistence` is to also send cache limiting headers. So pretty much all HTTP request going through PHP become non-cacheable.

And finally when configuring a cookie_lifetime, the absolute expiry date of the cookie will be re-set on every requests rendering the lifetime effectively infinite. This might be a desired use-case for some people, but it certainly should not be the default behavior as it is now impossible to have a fixed lifetime.

### Fix suggestion

If you agree with the idea, I'll submit a PR, and tests, to not call the persistence if the session is unchange. As simple as that:

```diff
diff --git src/SessionMiddleware.php src/SessionMiddleware.php
index ef77113..edafc5b 100644
--- src/SessionMiddleware.php
+++ src/SessionMiddleware.php
@@ -32,6 +32,11 @@ class SessionMiddleware implements MiddlewareInterface
     {
         $session = new LazySession($this->persistence, $request);
         $response = $handler->handle($request->withAttribute(self::SESSION_ATTRIBUTE, $session));
-        return $this->persistence->persistSession($session, $response);
+
+        if ($session->hasChanged()) {
+            $response = $this->persistence->persistSession($session, $response);
+        }
+
+        return $response;
     }
 }
```

### PR DESCRIPTION
**Pull request by PowerKiKi**
By doing so we avoid the following downsides:

- a (very) slight performance hit (on server CPU, client CPU and network
  transfer) for no gain at all
- pretty much all HTTP request going through PHP were becoming
  non-cacheable with default behavior of cache limiting headers
- when configuring a cookie_lifetime, the absolute expiry date of the
  cookie was re-set on every requests rendering the lifetime effectively
  infinite. That made it impossible to have a fixed lifetime

Fixes https://github.com/zendframework/zend-expressive-session/issues/36

Provide a narrative description of what you are trying to accomplish:

- [x] Are you fixing a bug?
  - [x] Detail how the bug is invoked currently.
  - [x] Detail the original, incorrect behavior.
  - [x] Detail the new, expected behavior.
  - [x] Base your feature on the `master` branch, and submit against that branch.
  - [x] Add a regression test that demonstrates the bug, and proves the fix.
  - [x] Add a `CHANGELOG.md` entry for the fix.

- [ ] Are you creating a new feature?
  - [ ] Why is the new feature needed? What purpose does it serve?
  - [ ] How will users use the new feature?
  - [ ] Base your feature on the `develop` branch, and submit against that branch.
  - [ ] Add only one feature per pull request; split multiple features over multiple pull requests
  - [ ] Add tests for the new feature.
  - [ ] Add documentation for the new feature.
  - [ ] Add a `CHANGELOG.md` entry for the new feature.

- [ ] Is this related to quality assurance?
  

- [ ] Is this related to documentation?

### CODE REVIEW

**Comment by PowerKiKi** [comment]
Coveralls report two lines that are uncovered by tests. I don't think those a re related to my changes. Are coverall reports flaky ?

**Review by geerteltink — APPROVED** [review]

**Comment by PowerKiKi** [comment]
@weierophinney you [expressed interest](https://github.com/zendframework/zend-expressive-session/issues/36#issuecomment-495688071) in this PR before. Could you please review it ?

**Comment by weierophinney** [comment]
@PowerKiKi Does this replace https://github.com/zendframework/zend-expressive-session/issues/36?

**Comment by PowerKiKi** [comment]
@weierophinney yes, this PR is an alternative to the solution suggested in https://github.com/zendframework/zend-expressive-session/issues/36. I'll close that issue once this PR is merged.

Do you think this can be merged as is ? or should we improve something ?

File: test/PhpSessionPersistenceTest.php [hunk 1]
```diff
@@ -188,8 +188,9 @@ public function testPersistSessionGeneratesCookieWithNewSessionIdIfSessionWasReg
      */
     public function testPersistSessionReturnsResponseWithSetCookieHeaderIfSessionCookiePresent()
     {
-        $request = $this->createSessionCookieRequest('use-this-id');
+        $request = $this->createSessionCookieRequest();
```
Comment (weierophinney) [inline test/PhpSessionPersistenceTest.php]: Why did you change this line? This test was originally partially validating that a new session identifier was created when persisting a session; the only change to this particular test should be the change you introduce on line 193.

**Review by weierophinney — DISMISSED** [review]
This looks good - except I have a question regarding one test change that I think needs to be addressed.

File: test/PhpSessionPersistenceTest.php (hunk 1 above)
Comment (PowerKiKi) [reply test/PhpSessionPersistenceTest.php]: I reverted that change and instead used `time()` to make sure that the session is indeed considered changed and a cookie is set.

If we don't do that, then the test will succeed on first run, but fail on any subsequent runs, because session saved on disk will be picked up and considered un-changed and thus a cookie will not be set, failing the test.

I believe a more proper solution is to always run all tests in a controlled environnement. `testNoMultipleEmptySessionFilesAreCreatedIfNoSessionCookiePresent` is one of a few tests that setup a custom `save_path` and take care of cleaning up that folder before running tests. I think this concept should be generalized to all tests and be done in `setUp()` instead of being copy/pasted as it is now. But that seemed out of scope, hence my quick fix. But if you agree on the principle I'd gladly submit a second PR for that.

Comment (weierophinney) [reply test/PhpSessionPersistenceTest.php]: :+1: to both: the rationale for the change, and an additional PR to move session cleaning to `setUp()`.

I'll get this merged soon!

**Comment by weierophinney** [comment]
Thanks, @PowerKiKi
